File: server/pkg/protocol/message_attribution_test.go
package protocol

import "testing"

func TestNormalizeMessageAttributionRequiresExplicitValidRanges(t *testing.T) {
	content := "alpha beta"
	sources, citations := NormalizeMessageAttribution(content,
		[]MessageSource{
			{ID: "docs", URL: "https://example.test/docs", Title: "Docs"},
			{ID: "", URL: "https://example.test/missing-id"},
			{ID: "docs", URL: "https://example.test/docs", Title: "Docs"},
		},
		[]MessageCitation{
			{SourceID: "docs", Start: 6, End: 10},
			{SourceID: "missing", Start: 0, End: 5},
			{SourceID: "docs", Start: 8, End: 20},
		},
	)
	if len(sources) != 1 || sources[0].URL != "https://example.test/docs" {
		t.Fatalf("sources = %#v", sources)
	}
	if len(citations) != 1 || citations[0] != (MessageCitation{SourceID: "docs", Start: 6, End: 10}) {
		t.Fatalf("citations = %#v", citations)
	}
}

func TestNormalizeMessageAttributionDropsAmbiguousSourceIDs(t *testing.T) {
	sources, citations := NormalizeMessageAttribution("done", []MessageSource{
		{ID: "same", URL: "https://example.test/a"},
		{ID: "same", URL: "https://example.test/b"},
	}, []MessageCitation{{SourceID: "same", Start: 0, End: 4}})
	if len(sources) != 0 || len(citations) != 0 {
		t.Fatalf("ambiguous attribution survived: sources=%#v citations=%#v", sources, citations)
	}
}

func TestNormalizeMessageAttributionUsesUTF8ByteBoundaries(t *testing.T) {
	source := []MessageSource{{ID: "unicode", URL: "https://example.test/unicode"}}
	_, citations := NormalizeMessageAttribution("猫", source, []MessageCitation{
		{SourceID: "unicode", Start: 0, End: 3},
		{SourceID: "unicode", Start: 0, End: 1},
	})
	if len(citations) != 1 || citations[0].End != 3 {
		t.Fatalf("citations = %#v", citations)
	}
}

File: packages/core/api/message-attribution.test.ts
import { describe, expect, it } from "vitest";
import { ChatMessageSchema } from "./schemas";
import { parseWSFrame } from "./ws-schema";

const source = {
  id: "docs",
  url: "https://example.test/docs",
  title: "Docs",
};
const citation = { source_id: "docs", start: 0, end: 4 };

describe("structured message attribution", () => {
  it("preserves explicit chat sources and never derives them from Markdown", () => {
    const plain = ChatMessageSchema.parse({
      id: "m1",
      chat_session_id: "s1",
      role: "assistant",
      content: "[Docs](https://example.test/docs)",
    });
    expect(plain.sources).toEqual([]);
    expect(plain.citations).toEqual([]);

    const attributed = ChatMessageSchema.parse({
      id: "m2",
      chat_session_id: "s1",
      role: "assistant",
      content: "done",
      sources: [source],
      citations: [citation],
    });
    expect(attributed.sources).toEqual([source]);
    expect(attributed.citations).toEqual([citation]);
  });

  it("keeps attribution on task-message and chat-done websocket frames", () => {
    for (const [type, payload] of [
      ["task:message", { task_id: "t1", seq: 1, type: "text", content: "done", sources: [source], citations: [citation] }],
      ["chat:done", { chat_session_id: "s1", task_id: "t1", message_id: "m1", content: "done", sources: [source], citations: [citation] }],
    ] as const) {
      const parsed = parseWSFrame({ type, payload });
      expect(parsed.kind).toBe("event");
      if (parsed.kind === "event") {
        expect((parsed.message.payload as { sources?: unknown }).sources).toEqual([source]);
        expect((parsed.message.payload as { citations?: unknown }).citations).toEqual([citation]);
      }
    }
  });
});
